sdk/storage/azure-storage-blob/tests/test_common_blob.py

```diff
@@ -3746,4 +3747,59 @@ def test_download_blob_no_decompress_chunks(self, **kwargs):
         result = blob.download_blob(decompress=False).readall()
         assert result == compressed_data
 
+    @pytest.mark.live_test_only
+    @BlobPreparer()
+    def test_dynamic_user_delegation_sas(self, **kwargs):
+        storage_account_name = kwargs.pop("storage_account_name")
+
+        token_credential = self.get_credential(BlobServiceClient)
+        service = BlobServiceClient(self.account_url(storage_account_name, "blob"), credential=token_credential)
+        container_name, blob_name = self.get_resource_name('oauthcontainer'), self.get_resource_name('oauthblob')
+        container = service.create_container(container_name)
+        blob = container.get_blob_client(blob_name)
+        blob.upload_blob(b"abc")
+
+        user_delegation_key = service.get_user_delegation_key(
+            key_start_time=datetime.utcnow(),
+            key_expiry_time=datetime.utcnow() + timedelta(hours=1)
+        )
+
+        request_headers = {
+            "foo$": "bar!",
+            "company": "msft",
+            "city": "redmond,atlanta,reston",
+        }
+
+        request_query_params = {
+            "hello$": "world!",
+            "abra": "cadabra",
+            "firstName": "john,Tim",
+        }
+
+        blob_token = self.generate_sas(
+            generate_blob_sas,
+            blob.account_name,
+            blob.container_name,
+            blob.blob_name,
+            permission=BlobSasPermissions(read=True),
+            expiry=datetime.utcnow() + timedelta(hours=1),
+            user_delegation_key=user_delegation_key,
+            request_headers=request_headers,
+            request_query_params=request_query_params
+        )
+
+        def callback(request):
+            request.http_request.headers["foo$"] = "world"
+            request.http_request.headers["company"] = "microsoft"
+            request.http_request.headers["city"] = "redmond,atlanta,reston"
+
+            parsed_url = urlparse(request.http_request.url)
+            query = urlencode(request_query_params)
+            url = urlunparse(parsed_url._replace(query=query))
```

Comment:
While technically works for this test, you should not replace any existing query params but instead append to the list of them. Look at `parse_qsl` to help.